### ISSUE TITLE
Support #[serde(rename_all)] for enum variants

### ISSUE THREAD

**Issue by Simon-Laux** [Aleph-Alpha/ts-rs#72]
Example:
 ```rs
use ts_rs::TS;
use serde::Serialize;

#[derive(Serialize, TS)]
#[serde(tag = "type")]
pub enum Thing {
    #[serde(rename_all = "camelCase")]
    A {
        display_name: Option<String>,
    },
    B (u8),
}
```
resulting warning:
```
warning: failed to parse serde attribute
  | 
  | #[serde(rename_all = "camelCase")]
  | 
  = note: ts-rs failed to parse this attribute. It will be ignored.
```

so the resulting variant does not have it's field-names renamed in camel casing.

see https://serde.rs/attributes.html

**Comment by phoneticsushi** [comment]
I encountered this independently.  Took a crack at it, but couldn't quite do it.

The immediate issue is in the `Parse` function called by the `TryFrom` implementation of the `SerdeEnumAttr` class:
https://github.com/Aleph-Alpha/ts-rs/blob/b79cb2a274df40da19686fee65f1142f5da4e272/macros/src/utils.rs#L99

...where both `Parse` and `TryFrom` are implemented by this macro:
https://github.com/Aleph-Alpha/ts-rs/blob/b79cb2a274df40da19686fee65f1142f5da4e272/macros/src/utils.rs#L15

...which is invoked here:
https://github.com/Aleph-Alpha/ts-rs/blob/b79cb2a274df40da19686fee65f1142f5da4e272/macros/src/attr/enum.rs#L83-L92

Expanding that invocation gives this - note the arms in the `match` block:
```rust
impl syn::parse::Parse for SerdeEnumAttr {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let mut out = SerdeEnumAttr::default();
        loop {
            let key: Ident = input.call(syn::ext::IdentExt::parse_any)?;
            match &*key.to_string() {
                "rename" => (out.0.rename = Some(parse_assign_str(input)?)),
                "rename_all" => (out.0.rename_all = Some(parse_assign_inflection(input)?)),
                "tag" => (out.0.tag = Some(parse_assign_str(input)?)),
                "content" => (out.0.content = Some(parse_assign_str(input)?)),
                "untagged" => (out.0.untagged = true),
                #[allow(unreachable_patterns)]
                _ => {
                    return Err(syn::Error::new((input.span()), {
                        let res = $crate::fmt::format(std::fmt::Arguments::new_v1(&[], &[]));
                        res
                    }))
                }
            }
            match input.is_empty() {
                true => break,
                false => {
                    input.parse::<$crate::token::Comma>()?;
                }
            }
        }
        Ok(out)
    }
}
```
Nothing is obviously wrong with this.

Let's add a couple `println`s to the `Parse` function:
```rust
                loop {
                    let key: Ident = $input.call(syn::ext::IdentExt::parse_any)?;
                    println!("The issue is here: key=<{}> input=<{}>", key.to_string(), $input);
                    match &*key.to_string() {
                        $($k => $e,)*
                        #[allow(unreachable_patterns)]
                        _ => {
                            println!("Match failed for <{}>", key.to_string());
                            syn_err!($input.span(); "unexpected attribute")
                        }
                    }
```
...and run @Simon-Laux's example.  The output is:
```
The issue is here: key=<rename_all> input=<= "camelCase">
Match failed for <rename_all>
warning: failed to parse serde attribute
  | 
  | #[serde(rename_all = "camelCase")]
  | 
  = note: ts-rs failed to parse this attribute. It will be ignored.
```
Great.

In summary,
- The printouts show that this code is being invoked on the variant's annotation
- "Match failed" shows it's an error selecting the arm of the match, not in the code that lies beneath
- This is the same code invoked when the attribute is applied to the enum, which is known to work
- "rename_all" doesn't work when the attribute is applied to the variant
- Other attributes work fine - replacing "rename_all" in the example with "rename" does in fact rename the variant itself

This is where my knowledge ends.  The key is clearly "rename_all".  The match statement clearly contains an arm for "rename_all".  Any ideas what I'm missing here?

**Comment by phoneticsushi** [comment]
I see what's missing now - the variant is being processed as a "SerdeFieldAttr" rather than a "SerdeEnumAttr":

```rust
                    match key_str.as_str() {
                        $($k => $e,)*
                        #[allow(unreachable_patterns)]
                        _ => {
                            let is_equal: bool = key_str.as_str().eq("rename_all");
                            println!("Error Case: [{}] | equal={}", key_str, is_equal);
                            println!("{}", std::any::type_name::<$i>());
                            syn_err!($input.span(); "unexpected attribute")
                        }
                    }
```
...yields:
```
Error Case: [rename_all] | equal=true
ts_rs_macros::attr::field::SerdeFieldAttr
warning: failed to parse serde attribute
  | 
  | #[serde(rename_all = "camelCase")]
  | 
  = note: ts-rs failed to parse this attribute. It will be ignored.
```
...which makes sense because `SerdeFieldAttr` lacks a case for "rename_all":
https://github.com/Aleph-Alpha/ts-rs/blob/b79cb2a274df40da19686fee65f1142f5da4e272/macros/src/attr/field.rs#L60-L78

**Comment by phoneticsushi** [comment]
I see what's missing now - the variant is being processed as a "SerdeFieldAttr" rather than a "SerdeEnumAttr":

```rust
                    match key_str.as_str() {
                        $($k => $e,)*
                        #[allow(unreachable_patterns)]
                        _ => {
                            let is_equal: bool = key_str.as_str().eq("rename_all");
                            println!("Error Case: [{}] | equal={}", key_str, is_equal);
                            println!("{}", std::any::type_name::<$i>());
                            syn_err!($input.span(); "unexpected attribute")
                        }
                    }
```
...yields:
```
Error Case: [rename_all] | equal=true
ts_rs_macros::attr::field::SerdeFieldAttr
warning: failed to parse serde attribute
  | 
  | #[serde(rename_all = "camelCase")]
  | 
  = note: ts-rs failed to parse this attribute. It will be ignored.
```
...which makes sense because `SerdeFieldAttr` lacks a case for "rename_all":
https://github.com/Aleph-Alpha/ts-rs/blob/b79cb2a274df40da19686fee65f1142f5da4e272/macros/src/attr/field.rs#L60-L78

**Comment by NyxCode** [comment]
fixed on master. Great work @phoneticsushi.
